readthedocs/proxito/exceptions.py

```diff
@@ -0,0 +1,174 @@
+from django.http import Http404
+from django.utils.translation import pgettext_lazy
+
+
+class ContextualizedHttp404(Http404):
+
+    """
+    Base class for contextualized HTTP 404 handling.
+
+    Subclasses may define their own template name,
+    HTTP status and object that was not found.
+
+    The contextualized exception is handled by proxito's 404 handler
+    """
+
+    template_name = "errors/404/base.html"
+    not_found_subject = pgettext_lazy(
+        "Names an object not found in a 404 error", "page"
+    )
+
+    def __init__(self, http_status=404, path_not_found=None, **kwargs):
+        """
+        Constructor that all subclasses should call.
+
+        :param kwargs: all kwargs are added as page context for rendering the 404 template
+        :param http_status: 404 view should respect this and set the HTTP status.
+        :param path_not_found: Inform the template and 404 view about a different path from
+                               request.path
+        """
+        self.http_status = http_status
+        self.path_not_found = path_not_found
+        self.kwargs = kwargs
+
+    def get_context(self):
+        c = {
+            "not_found_subject": self.not_found_subject,
+            "path_not_found": self.path_not_found,
+        }
+        c.update(self.kwargs)
+        return c
+
+
+class DomainDNSHttp404(ContextualizedHttp404):
+
+    """Raised if a DNS record points to us and we don't know the domain."""
+
+    template_name = "errors/404/dns.html"
+    not_found_subject = pgettext_lazy(
+        "Names an object not found in a 404 error", "matching DNS record"
+    )
+
+    def __init__(self, domain, **kwargs):
+        """
+        Raised when DNS for a custom domain is bad.
+
+        :param domain: The domain for which DNS is misconfigured.
+        :param kwargs:
+        """
+        kwargs["domain"] = domain
+        super().__init__(**kwargs)
+
+
+class ProjectHttp404(ContextualizedHttp404):
+
+    """
+    Raised if a domain did not resolve to a project.
+
+    This is currently used very broadly.
+    It indicates a number of reasons for the user.
+    """
+
+    template_name = "errors/404/no_project.html"
+    not_found_subject = pgettext_lazy(
+        "Names an object not found in a 404 error", "project"
+    )
+
+    def __init__(self, domain=None, **kwargs):
```

Comment:
Domain seems to always be passed here. This also doesn't require us to use `no domain` in the template, we should always have access to the domain.

```suggestion
    def __init__(self, domain, **kwargs):
```